singd/optim/optimizer.py

```diff
@@ -488,14 +488,20 @@ def _update_preconditioner(self, module: Module):
         H_K: StructuredMatrix = self.H_Ks.pop(module_name)
         H_C: StructuredMatrix = self.H_Cs.pop(module_name)
 
-        # un-scale `H_C = structure(C.T @ (grad_scale * g) @ (grad_scale * g).T @ C)`
+        # Define `grad_unscaling` to later un-scale
+        # `H_C = structure(C.T @ (grad_scale * g) @ (grad_scale * g).T @ C)`.
         prev_grad_scale = self._get_grad_scale(self.steps - 1)
         grad_scale = self._get_grad_scale(self.steps)
-        if grad_scale != 1.0 or prev_grad_scale != 1.0:
-            # In total we have to divide by `grad_scale ** 2`. The `H_C` computed
-            # in the backward pass was already divided by `prev_grad_scale` to avoid
-            # overflows. Here, we apply the remaining un-scaling
-            H_C.mul_(prev_grad_scale / grad_scale**2)
+        # In total we have to divide by `grad_scale ** 2`. The `H_C` computed
+        # in the backward pass was already divided by `prev_grad_scale` to avoid
+        # overflows. So we apply the remaining un-scaling. For increased
+        # numerical stability we do not scale `H_C` directly but instead
+        # include the un-scaling in the update of `m_K` and `m_C`.
+        grad_unscaling = (
+            prev_grad_scale / grad_scale**2
+            if grad_scale != 1.0 or prev_grad_scale != 1.0
+            else 1.0
+        )
```

Comment:
I remember we discussed this when I wrote that code: Does it make sense to refactor this into `grad_unscaling = prev_grad_scale / grad_scale ** 2` now? Without gradient scaling, both numbers will be `1.0` and their result will be `1.0` too.

Reply:
Yeah I don't see what could go wrong, so I'll refactor this.